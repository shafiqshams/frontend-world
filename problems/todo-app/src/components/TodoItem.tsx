import { Todo } from '../types/todo';

interface TodoItemProps {
   todo: Todo;
   toggleTodo: (todoId: number) => void;
   deleteTodo: (todoId: number) => void;
}

const TodoItem = ({ todo, toggleTodo, deleteTodo }: TodoItemProps) => {
   return (
      <div style={{ flex: 1, flexDirection: 'row', border: '2px' }}>
         <span
            data-testid={'todo-title'}
            onClick={() => toggleTodo(todo.id)}
            style={{ textDecoration: todo.completed ? 'line-through' : 'none', cursor: 'pointer' }}>
            {todo.title}
         </span>
         <button onClick={() => deleteTodo(todo.id)} data-testid='delete-btn'>Delete</button>
      </div>
   )
}

export default TodoItem;