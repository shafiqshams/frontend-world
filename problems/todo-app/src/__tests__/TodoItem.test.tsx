import { fireEvent, render } from "@testing-library/react"
import TodoItem from "../components/TodoItem"
import { Todo } from "../types/todo"
import '@testing-library/jest-dom';

describe('TodoItem', () => {

   const mockTodo: Todo = {
      id: 1,
      title: 'Unit test todo',
      completed: false
   };

   const toggleMock = jest.fn();
   const deleteMock = jest.fn();

   test('displays todo item when there is a todo', () => {

      //Arrange
      const { asFragment, getByText } = render(<TodoItem
         todo={mockTodo}
         toggleTodo={toggleMock}
         deleteTodo={deleteMock}
      />)

      //Assert
      expect(asFragment).toMatchSnapshot();
      expect(getByText('Unit test todo')).toBeInTheDocument();
   })

   test('calls toggleTodo when the text is clicked', () => {
      const { getByText } = render(<TodoItem todo={mockTodo} toggleTodo={toggleMock} deleteTodo={deleteMock} />);
      const todoText = getByText('Unit test todo');
      fireEvent.click(todoText);
      expect(toggleMock).toHaveBeenCalledWith(mockTodo.id);
   });

   test('calls deleteTodo when the delete button is clicked', () => {
      //Arrange
      const { getByTestId } = render(<TodoItem todo={mockTodo} toggleTodo={toggleMock} deleteTodo={deleteMock} />);
      const deleteButton = getByTestId('delete-btn');

      //Act
      fireEvent.click(deleteButton);

      //Assert
      expect(deleteMock).toHaveBeenCalledWith(mockTodo.id);
   });

   test('applies line-through style when todo is completed ', () => {
      //Arrange
      const { getByText } = render(<TodoItem todo={{ ...mockTodo, completed: true }} toggleTodo={toggleMock} deleteTodo={deleteMock} />);
      const todoText = getByText('Unit test todo');

      //No Action

      //Assert
      expect(todoText).toHaveStyle('text-decoration: line-through');
   });

   test('apply line-through style when todo is pressed', () => {
      //Arrange
      const { getByText, rerender } = render(<TodoItem todo={mockTodo} toggleTodo={toggleMock} deleteTodo={deleteMock} />);
      const todoText = getByText('Unit test todo');

      //Act
      fireEvent.click(todoText);


      //Assert
      expect(toggleMock).toHaveBeenCalledWith(mockTodo.id);
      // Simulate parent re-rendering the component with completed: true
      rerender(
         <TodoItem
            todo={{ ...mockTodo, completed: true }}
            toggleTodo={toggleMock}
            deleteTodo={deleteMock}
         />
      );
      expect(todoText).toHaveStyle('text-decoration: line-through');
   });
})