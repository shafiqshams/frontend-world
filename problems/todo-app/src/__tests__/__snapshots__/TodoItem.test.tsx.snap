// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem displays todo item when there is a todo 1`] = `[Function]`;
